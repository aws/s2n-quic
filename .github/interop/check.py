#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
#  SPDX-License-Identifier: Apache-2.0

import argparse
import json
import itertools
import sys

parser = argparse.ArgumentParser(description='Check interop reports.')
parser.add_argument('--required', type=str)
parser.add_argument('--regressions')
parser.add_argument('report')
args = parser.parse_args()

status = {'ok': True, 'failures': 0}
def fail(message):
    print(message)
    status['ok'] = False
    status['failures'] += 1

def format_required_report(report):
    outcome = {}
    result_idx = 0

    # find the version of s2n-quic to check
    s2n_quic = 's2n-quic'
    for impl in itertools.chain(report['clients'], report['servers']):
        # if we're testing a PR then use that name
        if impl.startswith('s2n-quic-pr'):
            s2n_quic = impl
            break

    for client in report['clients']:
        for server in report['servers']:
            result = report['results'][result_idx]
            result_idx += 1

            # we're only interested in s2n-quic results
            if client != s2n_quic and server != s2n_quic:
                continue

            for test in result:
                outcome.setdefault(test['name'], {})

                info = outcome[test['name']]
                info.setdefault(client, {'client': False, 'server': False})
                info.setdefault(server, {'client': False, 'server': False})

                success = test['result'] == 'succeeded'
                info[client]['client'] = success
                info[server]['server'] = success

    return outcome

with open(args.report) as f:
    result = json.load(f)

    if args.regressions and result['regression']:
        fail("A regression from main was detected")

    if args.required:
        with open(args.required) as r:
            required = json.load(r)

            actual = format_required_report(result)

            for test, impls in required.items():
                test_results = actual[test]
                for impl_name, endpoints in impls.items():
                    impl = test_results[impl_name]
                    for endpoint in endpoints:
                        if not impl[endpoint]:
                            fail("{} ({}) - {} was expected to pass but failed".format(impl_name, endpoint, test))

if not status['ok']:
    sys.exit(status['failures'])
