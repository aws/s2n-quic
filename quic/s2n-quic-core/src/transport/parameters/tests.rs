// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::*;
use bolero::check;
use s2n_codec::{assert_codec_round_trip_bytes, assert_codec_round_trip_value};

#[test]
#[cfg_attr(miri, ignore)] // This test is too expensive for miri to complete in a reasonable amount of time
fn round_trip() {
    check!().for_each(|input| {
        if input.is_empty() {
            return;
        }

        if input[0] > core::u8::MAX / 2 {
            assert_codec_round_trip_bytes!(ClientTransportParameters, input[1..]);
        } else {
            assert_codec_round_trip_bytes!(ServerTransportParameters, input[1..]);
        }
    });
}

macro_rules! default_transport_parameter_test {
    ($endpoint_params:ident) => {
        let default_value = $endpoint_params::default();

        #[cfg(not(miri))] // snapshot tests don't work on miri
        insta::assert_debug_snapshot!(
            concat!(stringify!($endpoint_params), "__default"),
            default_value
        );
        // Tests that a transport parameter will not be sent if it is set
        // to its default value defined in the rfc.
        let encoded_output: Vec<u8> =
            assert_codec_round_trip_value!($endpoint_params, default_value);
        let expected_output: Vec<u8> = vec![];
        assert_eq!(
            encoded_output, expected_output,
            "Default parameters should be empty"
        );
    };
}

#[test]
fn default_server_snapshot_test() {
    default_transport_parameter_test!(ServerTransportParameters);
}

#[test]
fn default_client_snapshot_test() {
    default_transport_parameter_test!(ClientTransportParameters);
}

fn server_transport_parameters() -> ServerTransportParameters {
    // pick a value that isn't the default for any of the params
    let integer_value = VarInt::from_u8(42);

    ServerTransportParameters {
        max_idle_timeout: integer_value.try_into().unwrap(),
        max_udp_payload_size: MaxUdpPayloadSize::new(1500u16).unwrap(),
        initial_max_data: integer_value.try_into().unwrap(),
        initial_max_stream_data_bidi_local: integer_value.try_into().unwrap(),
        initial_max_stream_data_bidi_remote: integer_value.try_into().unwrap(),
        initial_max_stream_data_uni: integer_value.try_into().unwrap(),
        initial_max_streams_bidi: integer_value.try_into().unwrap(),
        initial_max_streams_uni: integer_value.try_into().unwrap(),
        max_datagram_frame_size: MaxDatagramFrameSize::new(0u16).unwrap(),
        ack_delay_exponent: 2u8.try_into().unwrap(),
        max_ack_delay: integer_value.try_into().unwrap(),
        migration_support: MigrationSupport::Disabled,
        active_connection_id_limit: integer_value.try_into().unwrap(),
        original_destination_connection_id: Some([1, 2, 3, 4, 5, 6, 7, 8][..].try_into().unwrap()),
        stateless_reset_token: Some([2; 16].into()),
        preferred_address: Some(PreferredAddress {
            ipv4_address: Some(SocketAddressV4::new([127, 0, 0, 1], 1337)),
            ipv6_address: None,
            connection_id: [4, 5, 6, 7][..].try_into().unwrap(),
            stateless_reset_token: [1; 16].into(),
        }),
        initial_source_connection_id: Some([1, 2, 3, 4][..].try_into().unwrap()),
        retry_source_connection_id: Some([1, 2, 3, 4][..].try_into().unwrap()),
    }
}

#[test]
fn server_snapshot_test() {
    let value = server_transport_parameters();
    let encoded_output = assert_codec_round_trip_value!(ServerTransportParameters, value);

    #[cfg(not(miri))] // snapshot tests don't work on miri
    insta::assert_debug_snapshot!("server_snapshot_test", encoded_output);

    let _ = encoded_output;
}

fn client_transport_parameters() -> ClientTransportParameters {
    // pick a value that isn't the default for any of the params
    let integer_value = VarInt::from_u8(42);

    ClientTransportParameters {
        max_idle_timeout: integer_value.try_into().unwrap(),
        max_udp_payload_size: MaxUdpPayloadSize::new(1500u16).unwrap(),
        initial_max_data: integer_value.try_into().unwrap(),
        initial_max_stream_data_bidi_local: integer_value.try_into().unwrap(),
        initial_max_stream_data_bidi_remote: integer_value.try_into().unwrap(),
        initial_max_stream_data_uni: integer_value.try_into().unwrap(),
        initial_max_streams_bidi: integer_value.try_into().unwrap(),
        initial_max_streams_uni: integer_value.try_into().unwrap(),
        max_datagram_frame_size: MaxDatagramFrameSize::new(0u16).unwrap(),
        ack_delay_exponent: 2u8.try_into().unwrap(),
        max_ack_delay: integer_value.try_into().unwrap(),
        migration_support: MigrationSupport::Disabled,
        active_connection_id_limit: integer_value.try_into().unwrap(),
        original_destination_connection_id: Default::default(),
        stateless_reset_token: Default::default(),
        preferred_address: Default::default(),
        initial_source_connection_id: Some([1, 2, 3, 4][..].try_into().unwrap()),
        retry_source_connection_id: Default::default(),
    }
}

#[test]
fn client_snapshot_test() {
    let value = client_transport_parameters();
    let encoded_output = assert_codec_round_trip_value!(ClientTransportParameters, value);

    #[cfg(not(miri))] // snapshot tests don't work on miri
    insta::assert_debug_snapshot!("client_snapshot_test", encoded_output);

    let _ = encoded_output;
}

#[test]
fn load_server_limits() {
    let limits = crate::connection::limits::Limits::default();
    let mut params = ServerTransportParameters::default();
    params.load_limits(&limits);

    #[cfg(not(miri))]
    insta::assert_debug_snapshot!("load_server_limits", params);
}

#[test]
fn load_client_limits() {
    let limits = crate::connection::limits::Limits::default();
    let mut params = ClientTransportParameters::default();
    params.load_limits(&limits);

    #[cfg(not(miri))]
    insta::assert_debug_snapshot!("load_client_limits", params);
}

//= https://www.rfc-editor.org/rfc/rfc9000#section-7.4.2
//= type=test
//# An endpoint MUST ignore transport parameters that it does
//# not support.
#[test]
fn ignore_unknown_parameter() {
    use s2n_codec::EncoderBuffer;

    let value = client_transport_parameters();

    // Reserved parameters have tags of the form 31 * N + 27
    // We inject one at the end
    let mut buffer = vec![0; 32 * 1024];
    let mut encoder = EncoderBuffer::new(&mut buffer);

    encoder.encode(&value);

    let id1: TransportParameterId = VarInt::from_u16(31 * 2 + 27);
    encoder.encode(&id1);
    encoder.encode_with_len_prefix::<TransportParameterLength, _>(&());

    let (encoded, _) = encoder.split_off();
    let decoder = DecoderBuffer::new(encoded);
    let (decoded_params, remaining) =
        ClientTransportParameters::decode(decoder).expect("Decoding succeeds");
    assert_eq!(value, decoded_params);
    assert_eq!(0, remaining.len());
}

#[test]
fn compute_data_window_test() {
    assert_eq!(
        *compute_data_window(150, Duration::from_millis(10), 1),
        187_500
    );
    assert_eq!(
        *compute_data_window(150, Duration::from_millis(10), 2),
        375_000
    );
    assert_eq!(
        *compute_data_window(150, Duration::from_millis(100), 2),
        3_750_000
    );
    assert_eq!(
        *compute_data_window(1500, Duration::from_millis(100), 2),
        37_500_000
    );
}
