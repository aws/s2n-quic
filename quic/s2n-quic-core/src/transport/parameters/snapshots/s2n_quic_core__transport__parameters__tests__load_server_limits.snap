---
source: quic/s2n-quic-core/src/transport/parameters/tests.rs
assertion_line: 147
expression: params
---
TransportParameters {
    max_idle_timeout: MaxIdleTimeout(
        VarInt(
            30000,
        ),
    ),
    max_udp_payload_size: MaxUdpPayloadSize(
        VarInt(
            65527,
        ),
    ),
    initial_max_data: InitialMaxData(
        VarInt(
            3750000,
        ),
    ),
    initial_max_stream_data_bidi_local: InitialMaxStreamDataBidiLocal(
        VarInt(
            3750000,
        ),
    ),
    initial_max_stream_data_bidi_remote: InitialMaxStreamDataBidiRemote(
        VarInt(
            3750000,
        ),
    ),
    initial_max_stream_data_uni: InitialMaxStreamDataUni(
        VarInt(
            3750000,
        ),
    ),
    initial_max_streams_bidi: InitialMaxStreamsBidi(
        VarInt(
            100,
        ),
    ),
    initial_max_streams_uni: InitialMaxStreamsUni(
        VarInt(
            100,
        ),
    ),
    max_datagram_frame_size: MaxDatagramFrameSize(
        VarInt(
            0,
        ),
    ),
    ack_delay_exponent: AckDelayExponent(
        3,
    ),
    max_ack_delay: MaxAckDelay(
        VarInt(
            25,
        ),
    ),
    migration_support: Enabled,
    active_connection_id_limit: ActiveConnectionIdLimit(
        VarInt(
            2,
        ),
    ),
    original_destination_connection_id: None,
    stateless_reset_token: None,
    preferred_address: None,
    initial_source_connection_id: None,
    retry_source_connection_id: None,
}
